lib/semantics/resolve-names.cc

```diff
@@ -3767,35 +3784,33 @@ bool DeclarationVisitor::PassesLocalityChecks(
         "Assumed size array '%s' not allowed in a locality-spec"_err_en_US);
     return false;
   }
-  if (symbol.owner() == currScope()) {  // C1125 and C1126
-    SayAlreadyDeclared(name, symbol);
-    return false;
-  }
   // TODO: Check to see if the name can appear in a variable definition context
   return true;
 }
 
-Symbol *DeclarationVisitor::DeclareLocalEntity(const parser::Name &name) {
+Symbol *DeclarationVisitor::EnsureEnclosingEntity(const parser::Name &name) {
   Symbol *prev{FindSymbol(name)};
-  bool implicit{false};
   if (prev == nullptr) {
     // Declare the name as an object in the enclosing scope so that
     // the name can't be repurposed there later as something else.
     prev = &MakeSymbol(InclusiveScope(), name.source, Attrs{});
     ConvertToObjectEntity(*prev);
     ApplyImplicitRules(*prev);
-    implicit = true;
   }
+  return prev;
+}
+
+Symbol *DeclarationVisitor::DeclareLocalEntity(const parser::Name &name) {
+  Symbol *prev{EnsureEnclosingEntity(name)};
   if (!PassesLocalityChecks(name, *prev)) {
     return nullptr;
   }
   name.symbol = nullptr;
   Symbol &symbol{DeclareEntity<ObjectEntityDetails>(name, {})};
   if (auto *type{prev->GetType()}) {
-    if (implicit) {
-      ApplyImplicitRules(symbol);
-    } else {
-      symbol.SetType(*type);
+    symbol.SetType(*type);
+    if (prev->test(Symbol::Flag::Implicit)) {
+      symbol.set(Symbol::Flag::Implicit);
```

Comment:
You can write this as:
`symbol.set(Symbol::Flag::Implicit, prev->test(Symbol::Flag::Implicit))`